### PR TITLE
Eugene/task version

### PR DESCRIPTION
**Pull request by eugenejahn**
# TL;DR

- made the entities component become more generic that can support Workflow, Task, and others  
- added the inputs and outputs to task details page
- added the versions table in task details page 
- added the version details page 
- added task details link in node side panel 

Task Details Page:
https://localhost.demo.nuclyde.io:3000/projects/flytesnacks/domains/development/tasks/athena.athena.print_athena_schema?duration=all
image

Task Version Details Page:
https://localhost.demo.nuclyde.io:3000/projects/flytesnacks/domains/development/task/athena.athena.print_athena_schema/version/katrinathena6
image

Task Details Link in Executions Details Page
https://localhost.demo.nuclyde.io:3000/projects/flytesnacks/domains/development/executions/ei0ud0veie?duration=all
image

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyteconsole/issues/448

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import { WaitForData } from 'components/common/WaitForData';
+import { useProject } from 'components/hooks/useProjects';
+import { StaticGraphContainer } from 'components/Workflow/StaticGraphContainer';
+import { WorkflowId } from 'models/Workflow/types';
+import { entitySections } from 'components/Entities/constants';
+import { EntityDetailsHeader } from 'components/Entities/EntityDetailsHeader';
+import { EntityVersions } from 'components/Entities/EntityVersions';
+import { typeNameToEntityResource } from './constants';
+
+const useStyles = makeStyles((_theme: Theme) => ({
```
Comment (olga-union) [inline packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx:14]: nit: switch `_theme` with `theme` as it's used on line 20

Comment (olga-union) [inline packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx:14]: nit: switch `_theme` with `theme` as it's used on line 20

Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx]: tks! solved

File: packages/zapp/console/src/components/hooks/Entity/constants.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { listTasks } from 'models/Task/api';
+import { listWorkflows } from 'models/Workflow/api';
+import { Workflow } from 'models/Workflow/types';
+import { Task } from 'models/Task/types';
+import { PaginatedFetchFn } from '../types';
+
+type EntityStringMap = { [k in ResourceType]: string };
+
+export const entityStrings: EntityStringMap = {
+  [ResourceType.DATASET]: 'dataset',
```
Comment (ursucarina) [inline packages/zapp/console/src/components/hooks/Entity/constants.ts:11]: This seems to be duplicated in /components/Entities/constants.ts: entityStrings. can you use one or the other?

File: packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * A hook for fetching a paginated list of entity versions.
+ * @param scope
+ * @param config
+ */
+export function useEntityVersions(scope: IdentifierScope, config: RequestConfig) {
+  return usePagination<EntityType, IdentifierScope>(
+    { ...config, cacheItems: true, fetchArg: scope },
+    entityFunctions[(scope as Identifier)?.resourceType ?? Core.ResourceType.UNSPECIFIED]
+      .listEntity,
```
Comment (ursucarina) [inline packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts:20]: is listEntity always guaranteed to be present?

File: packages/zapp/console/src/models/__mocks__/taskData.ts [hunk 4]
```diff
@@ -12,8 +13,9 @@ const decodedClosure = Admin.TaskClosure.create(
 const taskId: (name: string, version: string) => Identifier = (name, version) => ({
   name,
   version,
-  project: 'flyte',
-  domain: 'development',
+  project: testProject,
```
Comment (ursucarina) [inline packages/zapp/console/src/models/__mocks__/taskData.ts:16]: no-op: nice!

File: packages/zapp/console/src/components/Entities/generators.ts [hunk 5]
```diff
@@ -40,3 +40,18 @@ export const backUrlGenerator: {
...
+  Routes.TaskDetails.makeUrl(project, domain, name);
+
+export const backToDetailUrlGenerator: {
+  [k in ResourceType]: (id: ResourceIdentifier) => string;
+} = {
+  [ResourceType.DATASET]: workflowDetailGenerator,
+  [ResourceType.LAUNCH_PLAN]: workflowDetailGenerator,
+  [ResourceType.TASK]: taskDetailGenerator,
+  [ResourceType.UNSPECIFIED]: workflowDetailGenerator,
+  [ResourceType.WORKFLOW]: workflowDetailGenerator,
```
Comment (ursucarina) [inline packages/zapp/console/src/components/Entities/generators.ts:56]: is using workflowDetailGenerator for all types correct here, or would it be better to either make it optional or create a dummy one that throws a not implemented error?

File: packages/zapp/console/src/components/Entities/EntityExecutions.tsx [hunk 6]
```diff
@@ -77,7 +79,7 @@ export const EntityExecutions: React.FC<EntityExecutionsProps> = ({
   return (
     <>
       <Typography className={styles.header} variant="h6">
-        All Executions in the Workflow
+        {t('allExecutionsChartTitle', entityStrings[id.resourceType])}
```
Comment (ursucarina) [inline packages/zapp/console/src/components/Entities/EntityExecutions.tsx:82]: nit: @anrusina to weigh in here. I like going through strings to do this, however piecing strings like this is not the most localization friendly for the future. What do you think?

**Comment by ursucarina** [comment]
Few comments regarding styling and code reuse

File: packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { IdentifierScope, Identifier } from 'models/Common/types';
+import { RequestConfig } from 'models/AdminEntity/types';
+import { listWorkflows } from 'models/Workflow/api';
```
[the remark this answers is not in the document]
Comment (eugenejahn) [reply packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts]: tks! solved

File: packages/zapp/console/src/components/hooks/Entity/constants.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { ResourceType } from 'models/Common/types';
+import { listTasks } from 'models/Task/api';
+import { listWorkflows } from 'models/Workflow/api';
+import { Workflow } from 'models/Workflow/types';
+import { Task } from 'models/Task/types';
+import { PaginatedFetchFn } from '../types';
```
[the remark this answers is not in the document]
Comment (eugenejahn) [reply packages/zapp/console/src/components/hooks/Entity/constants.ts]: tks! solved

File: packages/zapp/console/src/components/Executions/Tables/WorkflowVersionsTable.tsx [hunk 9]
```diff
@@ -28,20 +30,28 @@ interface WorkflowVersionRouteParams {
...
+  const { value: workflows, versionView, resourceType } = props;
   const state = useWorkflowExecutionsTableState();
   const commonStyles = useCommonStyles();
   const tableStyles = useExecutionTableStyles();
-  const { workflowVersion } = useParams<WorkflowVersionRouteParams>();
+  const { entityVersion } = useParams<WorkflowVersionRouteParams>();
 
   const columns = useWorkflowVersionsTableColumns();
 
   const retry = () => props.fetch();
```
[the remark this answers is not in the document]
Comment (eugenejahn) [reply packages/zapp/console/src/components/Executions/Tables/WorkflowVersionsTable.tsx]: tks! solved

File: packages/zapp/console/src/components/Entities/EntityDetailsHeader.tsx [hunk 10]
```diff
@@ -10,7 +10,7 @@ import * as React from 'react';
 import { Link } from 'react-router-dom';
 import { Routes } from 'routes/routes';
```
[the remark this answers is not in the document]
Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/EntityDetailsHeader.tsx]: tks! solved

File: packages/zapp/console/src/components/hooks/Entity/constants.ts (hunk 2 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/hooks/Entity/constants.ts]: removed this one!

File: packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts (hunk 3 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts]: added a try catch! tks

File: packages/zapp/console/src/components/Entities/generators.ts (hunk 5 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/generators.ts:62]: throw the error if not implmented! tks

File: packages/zapp/console/src/components/Entities/EntityExecutions.tsx (hunk 6 above)
Comment (anrusina) [reply packages/zapp/console/src/components/Entities/EntityExecutions.tsx:82]: Agree with Carina! As resourceType will always be in English, so when strings would be translated it would contain a mix of two languages, which is not ok. Also often when you translate sentence to a different language the words location is changed, for example "yet" word in Russian is moved between you and slept:
ENG: Have you slept **yet**?
RUS: Ты **уже** спал?

Please use patternKey for strings like this. Sample `label: t(patternKey('tableLabel', 'name')),` 
Here you will use as
```
{t(patternKey('allExecutionsChartTitle', entityStrings[id.resourceType]))}
```
In string file:
```
"allExecutionsChartTitle_": "Default string if entityStrings[id.resourceType] is null or undefined"
"allExecutionsChartTitle_Workflow": "String for Workflow"
"allExecutionsChartTitle_Tasks" : "String for Tasks"
```

Comment (anrusina) [reply packages/zapp/console/src/components/Entities/EntityExecutions.tsx:82]: I see that you probably used as example other string, found few other instances not created by you. Will cleanup them separately :)

File: packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import { useProject } from 'components/hooks/useProjects';
+import { StaticGraphContainer } from 'components/Workflow/StaticGraphContainer';
+import { WorkflowId } from 'models/Workflow/types';
+import { entitySections } from 'components/Entities/constants';
+import { EntityDetailsHeader } from 'components/Entities/EntityDetailsHeader';
+import { EntityVersions } from 'components/Entities/EntityVersions';
+import { typeNameToEntityResource } from './constants';
+
+const useStyles = makeStyles((theme: Theme) => ({
+  verionDetailsContatiner: {
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx:15]: nit: typo versionDetailsContatiner - missed **s** in version

File: packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    display: 'flex',
+    flexDirection: 'column',
+    flexWrap: 'nowrap',
+    overflow: 'hidden',
+    height: `calc(100vh - ${theme.spacing(1)}rem)`,
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/EntityVersionDetails.tsx:20]: nit: not sure we should use rem here, as most of our values in px and specing(1) assumes px

File: packages/zapp/console/src/components/Entities/strings.ts [hunk 13]
```diff
@@ -2,8 +2,8 @@ import { createLocalizedString } from 'basics/Locale';
 import { startCase } from 'lodash';
 
 const str = {
-  allExecutionsChartTitle: 'All Executions in the Workflow',
-  workflowVersionsTitle: 'Recent Workflow Versions',
+  allExecutionsChartTitle: (typeString: string) => `All Executions in the ${startCase(typeString)}`,
+  versionsTitle: (typeString: string) => `Recent ${startCase(typeString)} Versions`,
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/strings.ts:6]: change string as was talked in different comment, and update their usage.

File: packages/zapp/console/src/components/Entities/test/EntityDetails.test.tsx [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+  const createMocks = () => {
+    mockWorkflow = createMockWorkflow('MyWorkflow');
+    mockTask = createMockTask('MyTask');
+  };
+
+  const renderDetails = (id: ResourceIdentifier) => {
+    return render(
+      <MemoryRouter>
+        <EntityDetails id={id}></EntityDetails>
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/test/EntityDetails.test.tsx:33]: nit:  <EntityDetails id={id} /> - combine open and close tag

File: packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts (hunk 3 above)
Comment (anrusina) [reply packages/zapp/console/src/components/hooks/Entity/useEntityVersions.ts]: do we need to use ` xx.lisrEntity ?? someDefaultValue` or wrap it with try/catch?

**Review by anrusina — COMMENTED** [review]
Mainly nits.
The string need to be updates, gave a sample implementation to Carina's comment

**Comment by eugenejahn** [comment]
> Mainly nits. The string need to be updates, gave a sample implementation to Carina's comment

cool will fix it with next pr with the versions detail complete

File: packages/zapp/console/src/components/Entities/EntityDescription.tsx [hunk 15]
```diff
@@ -4,20 +4,83 @@ import classnames from 'classnames';
...
 
+const InputsAndOuputs: React.FC<{
+  id: ResourceIdentifier;
+}> = ({ id }) => {
+  const sort = {
+    key: executionSortFields.createdAt,
+    direction: SortDirection.DESCENDING,
+  };
+
+  const baseFilters = React.useMemo(
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/EntityDescription.tsx:43]: nit: I think this useMemo doesn't make sense, and a regular function should be enough. Mainly because `id` is an object - which means that any change to it (even if internal fields values haven't changed) will lead to pointer to id change and force useMemo to be regenerated.

File: packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+export interface EntityExecutionsProps {
+  id: ResourceIdentifier;
+}
+
+/** The tab/page content for viewing a workflow's executions */
+export const EntityVersionDetails: React.FC<EntityExecutionsProps> = ({ id }) => {
+  const styles = useStyles();
+
+  // TODO: need to be generic for supporting other type like workflow, etc.
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:36]: nit: Ideally `TODO:` need to have Task item mentioned - and new task to cover this work should be created

File: packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    <>
+      <Typography className={styles.header} variant="h3">
+        {t(patternKey('details', entityStrings[id.resourceType]))}
+      </Typography>
+      <div className={styles.divider} />
+      <WaitForData {...templateState}>
+        <div className={styles.table}>
+          {image && (
+            <Row title={t('imageFieldName')}>
+              {' '}
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:53]: Please remove `{' '}` here and below at this file. We should avoid using empty strings always. If you need some margin/paddings it's better to add a style for it.

Comment (anrusina) [reply packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:53]: And in EnvVarsTable.tsx file

File: packages/zapp/console/src/components/Entities/strings.ts [hunk 18]
```diff
@@ -1,15 +1,38 @@
...
 const str = {
-  allExecutionsChartTitle: 'All Executions in the Workflow',
-  workflowVersionsTitle: 'Recent Workflow Versions',
   viewAll: 'View All',
   schedulesHeader: 'Schedules',
   collapseButton: (showAll: boolean) => (showAll ? 'Collapse' : 'Expand'),
-  launchStrings: (typeString: string) => `Launch ${startCase(typeString)}`,
-  noDescription: (typeString: string) => `This ${typeString} has no description.`,
-  noSchedules: (typeString: string) => `This ${typeString} has no schedules.`,
+  launchStrings_workflow: launchStrings('Workflow'),
```
Comment (anrusina) [inline packages/zapp/console/src/components/Entities/strings.ts:15]: nit: in real string files ready for localization we would never have functions inside, so ideally `This Workflow has no schedules.` should be used here and below instead of `launchStrings('Workflow')`.
Mainly note for future, as we anyway would need tp go though all our files before we will be ready to add real localization support.

File: packages/zapp/console/src/components/Executions/ExecutionDetails/NodeExecutionDetailsPanelContent.tsx [hunk 19]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import { useEffect, useMemo, useRef, useState } from 'react';
-import { IconButton, Typography, Tab, Tabs } from '@material-ui/core';
+import { IconButton, Typography, Tab, Tabs, Link } from '@material-ui/core';
```
Comment (anrusina) [inline packages/zapp/console/src/components/Executions/ExecutionDetails/NodeExecutionDetailsPanelContent.tsx:3]: nit: Link is not needed, please remove

**Review by anrusina — COMMENTED** [review]
Mainly nits, One or two comments required fix. 

It also has a big drop in test coverage.. Can you please work on increasing t for files you worked with? You can see missed checks in PR marked with notifications as:
Screen Shot 2022-05-17 at 10 43 27 AM

Or you can run `yarn test-coverage` and than open coverage info in browser. For me it's `file:///Users/narusina/prj/flyte/flyteconsole/.coverage/index.html` - You can find you relative path.

Otherwise LGTM

File: packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx (hunk 17 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:53]: eslint auto did that :( yeah I will remove it

File: packages/zapp/console/src/components/Entities/strings.ts (hunk 18 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/strings.ts:15]: ohhh. I see now. that makes sense will fix it

File: packages/zapp/console/src/components/Executions/ExecutionDetails/NodeExecutionDetailsPanelContent.tsx (hunk 19 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/Executions/ExecutionDetails/NodeExecutionDetailsPanelContent.tsx:3]: got it thanks

File: packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx (hunk 16 above)
Comment (eugenejahn) [reply packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:36]: or should I change to NOTE: ?

Comment (anrusina) [reply packages/zapp/console/src/components/Entities/VersionDetails/EntityVersionDetails.tsx:36]: Maybe change to NOTE: here

**Review by anrusina — APPROVED** [review]
LGTM, Nice Work

# 🚀
